ruby/ql/lib/codeql/ruby/ast/Literal.qll

```diff
@@ -129,13 +167,32 @@ class RationalLiteral extends NumericLiteral, TRationalLiteral {
 
   RationalLiteral() { this = TRationalLiteral(g) }
 
-  final override string getValueText() { result = g.getChild().(Ruby::Token).getValue() + "r" }
+  final override ConstantValue::ConstantRationalValue getConstantValue() {
+    exists(int numerator, int denominator |
+      isRationalValue(g, numerator, denominator) and
+      result.isRational(numerator, denominator)
+    )
+  }
 
-  final override string toString() { result = this.getValueText() }
+  final override string toString() { result = g.getChild().(Ruby::Token).getValue() + "r" }
 
   final override string getAPrimaryQlClass() { result = "RationalLiteral" }
 }
 
+private float getComplexValue(Ruby::Complex c) {
+  exists(string s |
+    s = c.getValue() and
+    result = s.prefix(s.length() - 1).toFloat()
+  )
+}
+
+private class RequiredComplexConstantValue extends RequiredConstantValue {
+  override predicate requiredComplex(float real, float imaginary) {
+    real = 0 and
+    imaginary = getComplexValue(_)
+  }
+}
```

Comment:
This doesn't appear to support complex numbers like `1+2i` or `3-4i`. Is it worth adding that?

Reply:
That would have to be done in `BinaryOperationCfgNode`. We could add constant folding for complex and rational numbers there, but I think that should be done as follow-up work.

Reply:
Unless I'm missing something, I don't think so? `1+2i` is parsed as a single `ComplexLiteral` value, not an application of `+` to two separate literals.

Reply:
(Totally happy if we do this as follow-up work, but I think this is the right part of the code to be changing)

Reply:
👍 Let's do it as follow-up.